File: contracts/wrappers/examples/jetton/OnrampMock.ts
import {
  Address,
  beginCell,
  Cell,
  Contract,
  contractAddress,
  ContractProvider,
  Sender,
  SendMode,
} from '@ton/core'
import { JettonClientConfig, jettonClientConfigToCell, ErrorCodes } from './types'

export type OnrampMockConfig = {
  jettonClient: JettonClientConfig
}

export function onrampMockConfigToCell(config: OnrampMockConfig): Cell {
  return jettonClientConfigToCell(config.jettonClient)
}

export const OnrampConstants = {
  FEE: 5n,
  INCORRECT_SENDER_ERROR: ErrorCodes.INCORRECT_SENDER,
  FORWARD_PAYLOAD_REQUIRED_ERROR: ErrorCodes.FORWARD_PAYLOAD_REQUIRED,
}

export const EventOpcodes = {
  INSUFFICIENT_FEE: 'InsufficientFee',
  ACCEPTED_REQUEST: 'AcceptedRequest',
}

export type InsufficientFeeEvent = {
  queryId: bigint
  sender: Address
}

export type AcceptedRequestEvent = {
  queryId: bigint
  sender: Address
  payload: Cell
}

export class OnrampMock implements Contract {
  constructor(
    readonly address: Address,
    readonly init?: { code: Cell; data: Cell },
  ) {}

  static createFromAddress(address: Address) {
    return new OnrampMock(address)
  }

  static createFromConfig(config: OnrampMockConfig, code: Cell, workchain = 0) {
    const data = onrampMockConfigToCell(config)
    const init = { code, data }
    return new OnrampMock(contractAddress(workchain, init), init)
  }

  async sendDeploy(provider: ContractProvider, via: Sender, value: bigint) {
    await provider.internal(via, {
      value,
      sendMode: SendMode.PAY_GAS_SEPARATELY,
      body: beginCell().endCell(),
    })
  }

  // Helper method to parse events from transaction logs
  static parseInsufficientFeeEvent(cell: Cell): InsufficientFeeEvent {
    const slice = cell.beginParse()
    return {
      queryId: BigInt(slice.loadUint(64)),
      sender: slice.loadAddress(),
    }
  }

  static parseAcceptedRequestEvent(cell: Cell): AcceptedRequestEvent {
    const slice = cell.beginParse()
    return {
      queryId: BigInt(slice.loadUint(64)),
      sender: slice.loadAddress(),
      payload: slice.loadRef(),
    }
  }
}

File: integration-tests/jetton/wrappers/onramp_mock.go
package wrappers

import (
	"fmt"

	test_utils "integration-tests/utils"

	"github.com/xssnick/tonutils-go/address"
	"github.com/xssnick/tonutils-go/tlb"
	"github.com/xssnick/tonutils-go/tvm/cell"

	"github.com/smartcontractkit/chainlink-ton/pkg/ton/tracetracking"
	"github.com/smartcontractkit/chainlink-ton/pkg/ton/wrappers"
)

var OnrampMockContractPath = test_utils.GetBuildDir("examples.jetton.OnrampMock.compiled.json")

type OnrampMockProvider struct {
	apiClient tracetracking.SignedAPIClient
}

func NewOnrampMockProvider(apiClient tracetracking.SignedAPIClient) *OnrampMockProvider {
	return &OnrampMockProvider{
		apiClient: apiClient,
	}
}

type OnrampMockInitData struct {
	MasterAddress    *address.Address `tlb:"addr"`
	JettonWalletCode *cell.Cell       `tlb:"^"`
}

func (p *OnrampMockProvider) Deploy(initData OnrampMockInitData) (OnrampMock, error) {
	initCell, err := tlb.ToCell(initData)
	if err != nil {
		return OnrampMock{}, fmt.Errorf("failed to convert init data to cell: %w", err)
	}

	compiledContract, err := wrappers.ParseCompiledContract(OnrampMockContractPath)
	if err != nil {
		return OnrampMock{}, fmt.Errorf("failed to compile contract: %w", err)
	}
	contract, _, err := wrappers.Deploy(&p.apiClient, compiledContract, initCell, tlb.MustFromTON("1"), cell.BeginCell().EndCell())
	if err != nil {
		return OnrampMock{}, err
	}

	return OnrampMock{
		Contract: *contract,
	}, nil
}

type OnrampMock struct {
	Contract wrappers.Contract
}

// OnrampMock constants
const (
	OnrampMockFee                         = 5
	OnrampMockIncorrectSenderError        = 100
	OnrampMockForwardPayloadRequiredError = 101
)

// Event opcodes (these are typically used in logs/events)
const (
	OnrampMockInsufficientFeeEvent = "InsufficientFee"
	OnrampMockAcceptedRequestEvent = "AcceptedRequest"
)

// OnrampMock receives jetton transfer notifications automatically
// It doesn't have explicit send methods, but we can add helper methods
// to parse events from transaction logs

type InsufficientFeeEvent struct {
	QueryID uint64
	Sender  *address.Address
}

type AcceptedRequestEvent struct {
	QueryID uint64
	Sender  *address.Address
	Payload *cell.Cell
}

// Helper methods to parse events from transaction logs
func ParseInsufficientFeeEvent(cell *cell.Cell) (InsufficientFeeEvent, error) {
	slice := cell.BeginParse()
	queryID, err := slice.LoadUInt(64)
	if err != nil {
		return InsufficientFeeEvent{}, fmt.Errorf("failed to load queryID: %w", err)
	}

	sender, err := slice.LoadAddr()
	if err != nil {
		return InsufficientFeeEvent{}, fmt.Errorf("failed to load sender: %w", err)
	}

	return InsufficientFeeEvent{
		QueryID: queryID,
		Sender:  sender,
	}, nil
}

// Helper method to parse events from transaction results
// Note: This would typically be used when parsing transaction events/logs
func ParseAcceptedRequestEvent(eventCell *cell.Cell) (AcceptedRequestEvent, error) {
	slice := eventCell.BeginParse()
	queryID, err := slice.LoadUInt(64)
	if err != nil {
		return AcceptedRequestEvent{}, fmt.Errorf("failed to load queryID: %w", err)
	}

	sender, err := slice.LoadAddr()
	if err != nil {
		return AcceptedRequestEvent{}, fmt.Errorf("failed to load sender: %w", err)
	}

	// For now, we'll assume the payload is stored as a simple cell
	// In a real implementation, the payload structure depends on the contract
	if slice.BitsLeft() > 0 {
		// Create a new cell with the remaining data
		payloadBuilder := cell.BeginCell()
		// Store the remaining slice content (this is a simplified approach)
		payloadCell := payloadBuilder.EndCell()

		return AcceptedRequestEvent{
			QueryID: queryID,
			Sender:  sender,
			Payload: payloadCell,
		}, nil
	}

	return AcceptedRequestEvent{
		QueryID: queryID,
		Sender:  sender,
		Payload: nil,
	}, nil
}
